geode-core/src/main/java/org/apache/geode/internal/admin/remote/ShutdownAllRequest.java

```diff
@@ -259,7 +260,11 @@ public void process(DistributionMessage msg) {
       }
       if (msg instanceof ShutdownAllResponse) {
         if (((ShutdownAllResponse) msg).isToShutDown()) {
-          logger.debug("{} adding {} to result set {}", this, msg.getSender(), this.results);
+          if (logger.isDebugEnabled()) {
+            Set<DistributedMember> copySet = new TreeSet<>(results);
```

Comment:
Doesn't this have the same risk of throwing ConcurrentModificationException?
I think this issue is with doing an iteration of the synchronizedSet "results" while not synchronized on it.
The implicit toString call in the logging code did an iteration.
Now I think an implicit iteration is done over "results" when it is passed to the copy constructor "TreeSet(Collection)".
I think you want to hold a synchronize on "results" while calling the copy constructor. Something like this:
Set<DistributedMember> copySet;
synchronized (results) {
  copySet = new TreeSet<>(results);
}

Reply:
Even better; just add "synchronized (this.results) {}" inside the isDebugEnabled check and around the logger.debug line. No need to make a copy of the set.